apps/api/src/sub_refresh_job.rs

```diff
@@ -0,0 +1,145 @@
+use std::sync::Arc;
+
+use shared::database::{
+	duration::DurationUnit,
+	entitlement::{EntitlementEdge, EntitlementEdgeId, EntitlementEdgeKind, EntitlementEdgeManagedBy},
+	product::{
+		subscription::{SubscriptionId, SubscriptionPeriod},
+		SubscriptionBenefit, SubscriptionBenefitCondition,
+	},
+	queries::filter,
+};
+
+use crate::{
+	global::Global,
+	http::error::ApiError,
+	transactions::{TransactionError, TransactionResult, TransactionSession},
+};
+
+pub fn sub_age_days(periods: &[SubscriptionPeriod]) -> isize {
+	sub_age(periods).0
+}
+
+pub fn sub_age(periods: &[SubscriptionPeriod]) -> (isize, isize) {
+	periods
+		.iter()
+		.map(|p| {
+			let diff = date_component::date_component::calculate(&p.start, &p.end);
+			(diff.interval_days, diff.month)
+		})
+		.fold((0, 0), |(days, month), diff| (days + diff.0, month + diff.1))
+}
+
+/// Grants entitlements for a subscription.
+pub async fn refresh_entitlements(
+	tx: &mut TransactionSession<'_, ApiError>,
+	subscription_id: &SubscriptionId,
+	benefits: &[SubscriptionBenefit],
+) -> TransactionResult<(), ApiError> {
+	// first revoke all entitlements
+	revoke_entitlements(tx, subscription_id).await?;
```

Comment:
I do not like this. Rather compute exactly what needs to change and then change that.